File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/assert-reconstruction-integrity.py
#!/usr/bin/env python3
"""Tableau-local completion gate for reconstructed dashboard residue."""

from __future__ import annotations

import argparse
import json
import re
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


FAMILY_MAP = {
    "bar": "bar",
    "bar-chart": "bar",
    "column": "bar",
    "column-chart": "bar",
    "line": "line",
    "line-chart": "line",
    "sparkline": "line",
    "area": "area",
    "area-chart": "area",
    "combo": "combo",
    "combo-chart": "combo",
    "dual-axis": "combo",
    "scatter": "scatter",
    "scatter-chart": "scatter",
    "bubble": "scatter",
    "pie": "pie",
    "pie-chart": "pie",
    "donut": "pie",
    "donut-chart": "pie",
    "kpi": "kpi",
    "kpi-chart": "kpi",
    "single-value": "kpi",
    "big-number": "kpi",
    "map": "map",
    "region-map": "map",
    "point-map": "map",
    "table": "table",
    "pivot-table": "table",
    "pivot": "table",
    "crosstab": "table",
    "text-table": "table",
    "grid": "table",
}


def read_json(path: Path) -> Any:
    try:
        return json.loads(path.read_text(encoding="utf-8-sig"))
    except json.JSONDecodeError as exc:
        raise ValueError(f"{path} is not valid JSON: {exc}") from exc


def norm(value: Any) -> str:
    return re.sub(r"[^a-z0-9]", "", str(value or "").strip().lower())


def workbook_elements(readback: Any) -> list[dict[str, Any]]:
    if not isinstance(readback, dict):
        return []
    document = readback.get("document")
    document = document if isinstance(document, dict) else readback
    elements = document.get("elements")
    if isinstance(elements, list):
        return [element for element in elements if isinstance(element, dict)]
    return [
        element
        for page in document.get("pages", [])
        if isinstance(page, dict)
        for element in page.get("elements", [])
        if isinstance(element, dict)
    ]


def evaluate(workdir: Path) -> dict[str, Any]:
    control_blockers: list[dict[str, Any]] = []
    census_paths = sorted(workdir.glob("*-controls-coverage.json"))
    census_path = census_paths[0] if census_paths else None
    if census_path:
        census = read_json(census_path)
        rows = census.get("detail") if isinstance(census, dict) else None
        if not isinstance(rows, list):
            raise ValueError(f"{census_path} is malformed (expected detail array)")
        waivers_path = workdir / "controls-waivers.json"
        waivers = read_json(waivers_path) if waivers_path.exists() else []
        waivers = waivers.get("waivers", []) if isinstance(waivers, dict) else waivers
        valid_waivers = [
            waiver
            for waiver in waivers
            if isinstance(waiver, dict)
            and str(waiver.get("control") or waiver.get("name") or "").strip()
            and str(waiver.get("reason") or "").strip()
        ]
        for row in rows:
            if not isinstance(row, dict) or row.get("status") not in {
                "needs-wiring",
                "needs-materialization",
            }:
                continue
            kind = str(row.get("kind") or "")
            name = str(row.get("name") or "")
            waived = any(
                norm(waiver.get("control") or waiver.get("name"))
                in {norm(name), norm(f"{kind}:{name}")}
                for waiver in valid_waivers
            )
            if not waived:
                control_blockers.append(
                    {"kind": kind, "name": name, "status": row.get("status")}
                )

    kind_blockers: list[dict[str, Any]] = []
    png_path = workdir / "png-read.json"
    readback_path = workdir / "wb-readback.json"
    renames_path = workdir / "layout-renames.json"
    if png_path.exists() and readback_path.exists() and renames_path.exists():
        png = read_json(png_path)
        readback = read_json(readback_path)
        renames = read_json(renames_path)
        if (
            isinstance(png, dict)
            and png.get("verified") is not False
            and isinstance(png.get("tiles"), list)
            and isinstance(renames, dict)
        ):
            elements_by_name: dict[str, list[dict[str, Any]]] = {}
            for element in workbook_elements(readback):
                if element.get("visibleAsSource") is False:
                    continue
                built_family = FAMILY_MAP.get(
                    str(element.get("kind") or "").strip().lower()
                )
                if not built_family:
                    continue
                name = element.get("name")
                name = name.get("text") if isinstance(name, dict) else name
                name = name or element.get("title") or element.get("id")
                key = norm(name)
                if key:
                    elements_by_name.setdefault(key, []).append(
                        {
                            "id": element.get("id"),
                            "kind": element.get("kind"),
                            "family": built_family,
                        }
                    )
            rename_by_source = {
                norm(source_name): built_name
                for source_name, built_name in renames.items()
            }
            kind_waivers = [
                waiver
                for waiver in png.get("kind_waivers", [])
                if isinstance(waiver, dict)
                and str(waiver.get("tile") or "").strip()
                and str(waiver.get("reason") or "").strip()
            ]
            for tile in png["tiles"]:
                if not isinstance(tile, dict):
                    continue
                expected = FAMILY_MAP.get(
                    str(tile.get("kind") or tile.get("chart_kind") or "")
                    .strip()
                    .lower()
                )
                source_name = str(tile.get("title") or "")
                built_name = rename_by_source.get(norm(source_name))
                if not expected or not built_name:
                    continue
                actuals = elements_by_name.get(norm(built_name), [])
                if not actuals or any(
                    element["family"] == expected for element in actuals
                ):
                    continue
                if any(
                    norm(waiver.get("tile")) == norm(source_name)
                    for waiver in kind_waivers
                ):
                    continue
                kind_blockers.append(
                    {
                        "source_tile": source_name,
                        "renamed_element": built_name,
                        "expected_family": expected,
                        "built_families": sorted(
                            {element["family"] for element in actuals}
                        ),
                        "built_kinds": sorted(
                            {str(element["kind"]) for element in actuals}
                        ),
                    }
                )

    return {
        "version": 1,
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "status": (
            "PASS" if not control_blockers and not kind_blockers else "FAIL"
        ),
        "controls_census": census_path.name if census_path else None,
        "unresolved_controls": control_blockers,
        "renamed_chart_family_mismatches": kind_blockers,
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--workdir", required=True, type=Path)
    args = parser.parse_args()
    try:
        result = evaluate(args.workdir)
    except (OSError, ValueError) as exc:
        print(f"FATAL: {exc}", file=sys.stderr)
        return 1

    output_path = args.workdir / "reconstruction-integrity.json"
    output_path.write_text(
        json.dumps(result, indent=2, ensure_ascii=False) + "\n", encoding="utf-8"
    )
    for blocker in result["unresolved_controls"]:
        print(
            f"[FAIL] unfinished control: {blocker['kind']}:{blocker['name']} "
            f"({blocker['status']})",
            file=sys.stderr,
        )
    for blocker in result["renamed_chart_family_mismatches"]:
        print(
            f"[FAIL] renamed chart family: {blocker['source_tile']!r} -> "
            f"{blocker['renamed_element']!r}: expected "
            f"{blocker['expected_family']}, built "
            f"{'/'.join(blocker['built_families'])}",
            file=sys.stderr,
        )
    if result["status"] == "PASS":
        print(
            "[OK] reconstruction integrity: controls terminal, renamed chart "
            "families faithful or explicitly waived"
        )
        print(f"wrote {output_path}")
        return 0
    print(f"wrote {output_path}", file=sys.stderr)
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/assert-reconstruction-integrity.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Tableau-local completion gate for manual/reconstructed dashboard work:
#   1. `needs-wiring` / `needs-materialization` controls are unfinished work,
#      not terminal coverage decisions. They block unless controls-waivers.json
#      explicitly accepts the scope cut with a reason.
#   2. layout-renames.json cannot hide a chart-family substitution from the
#      verified png-read.json → live wb-readback comparison. A deliberate
#      substitution needs a reasoned png-read kind_waivers entry.
#
# Kept local rather than folded into the vendored shared assert-phase6-ran.rb:
# reconstruction rename sidecars and Tableau controls-coverage statuses are
# plugin-specific.

require 'json'
require 'optparse'
require 'time'
require_relative 'lib/code_rep'

opts = {}
OptionParser.new do |parser|
  parser.on('--workdir DIR') { |value| opts[:workdir] = File.expand_path(value) }
end.parse!
abort('usage: assert-reconstruction-integrity.rb --workdir DIR') unless opts[:workdir]

workdir = opts[:workdir]
norm = ->(value) { value.to_s.strip.downcase.gsub(/[^a-z0-9]/, '') }
read_json = lambda do |path|
  JSON.parse(File.read(path))
rescue JSON::ParserError => e
  abort("FATAL: #{path} is not valid JSON: #{e.message}")
end

control_blockers = []
control_census_path = Dir[File.join(workdir, '*-controls-coverage.json')].min
if control_census_path
  census = read_json.call(control_census_path)
  rows = census.is_a?(Hash) ? census['detail'] : nil
  abort("FATAL: #{control_census_path} is malformed (expected detail array)") unless rows.is_a?(Array)

  waivers_path = File.join(workdir, 'controls-waivers.json')
  waivers = File.exist?(waivers_path) ? read_json.call(waivers_path) : []
  waivers = waivers['waivers'] if waivers.is_a?(Hash)
  valid_waivers = Array(waivers).select do |waiver|
    waiver.is_a?(Hash) &&
      !(waiver['control'] || waiver['name']).to_s.strip.empty? &&
      !waiver['reason'].to_s.strip.empty?
  end

  rows.each do |row|
    next unless row.is_a?(Hash) && %w[needs-wiring needs-materialization].include?(row['status'].to_s)
    kind = row['kind'].to_s
    name = row['name'].to_s
    waived = valid_waivers.any? do |waiver|
      key = norm.call(waiver['control'] || waiver['name'])
      key == norm.call(name) || key == norm.call("#{kind}:#{name}")
    end
    control_blockers << {
      'kind' => kind,
      'name' => name,
      'status' => row['status']
    } unless waived
  end
end

family_map = {
  'bar' => 'bar', 'bar-chart' => 'bar', 'column' => 'bar', 'column-chart' => 'bar',
  'line' => 'line', 'line-chart' => 'line', 'sparkline' => 'line',
  'area' => 'area', 'area-chart' => 'area',
  'combo' => 'combo', 'combo-chart' => 'combo', 'dual-axis' => 'combo',
  'scatter' => 'scatter', 'scatter-chart' => 'scatter', 'bubble' => 'scatter',
  'pie' => 'pie', 'pie-chart' => 'pie', 'donut' => 'pie', 'donut-chart' => 'pie',
  'kpi' => 'kpi', 'kpi-chart' => 'kpi', 'single-value' => 'kpi', 'big-number' => 'kpi',
  'map' => 'map', 'region-map' => 'map', 'point-map' => 'map',
  'table' => 'table', 'pivot-table' => 'table', 'pivot' => 'table',
  'crosstab' => 'table', 'text-table' => 'table', 'grid' => 'table'
}.freeze
family = ->(kind) { family_map[kind.to_s.strip.downcase] }

kind_blockers = []
png_path = File.join(workdir, 'png-read.json')
readback_path = File.join(workdir, 'wb-readback.json')
renames_path = File.join(workdir, 'layout-renames.json')
if File.exist?(png_path) && File.exist?(readback_path) && File.exist?(renames_path)
  png = read_json.call(png_path)
  readback = read_json.call(readback_path)
  renames = read_json.call(renames_path)
  if png.is_a?(Hash) && png['verified'] != false && png['tiles'].is_a?(Array) && renames.is_a?(Hash)
    elements_by_name = Hash.new { |hash, key| hash[key] = [] }
    Sigma::CodeRep.workbook_elements(readback).each do |element|
      next unless element.is_a?(Hash) && element['visibleAsSource'] != false
      built_family = family.call(element['kind'])
      next unless built_family
      name = element['name'].is_a?(Hash) ? element.dig('name', 'text') : element['name']
      name = element['title'] || element['id'] if name.to_s.strip.empty?
      key = norm.call(name)
      elements_by_name[key] << {
        'id' => element['id'],
        'kind' => element['kind'],
        'family' => built_family
      } unless key.empty?
    end

    rename_by_source = renames.each_with_object({}) do |(source_name, built_name), index|
      index[norm.call(source_name)] = built_name
    end
    kind_waivers = Array(png['kind_waivers']).select do |waiver|
      waiver.is_a?(Hash) && !waiver['tile'].to_s.strip.empty? && !waiver['reason'].to_s.strip.empty?
    end

    png['tiles'].each do |tile|
      next unless tile.is_a?(Hash)
      expected = family.call(tile['kind'] || tile['chart_kind'])
      source_name = tile['title'].to_s
      built_name = rename_by_source[norm.call(source_name)]
      next unless expected && built_name
      actuals = elements_by_name[norm.call(built_name)]
      next if actuals.empty? || actuals.any? { |element| element['family'] == expected }
      waived = kind_waivers.any? { |waiver| norm.call(waiver['tile']) == norm.call(source_name) }
      next if waived
      kind_blockers << {
        'source_tile' => source_name,
        'renamed_element' => built_name,
        'expected_family' => expected,
        'built_families' => actuals.map { |element| element['family'] }.uniq,
        'built_kinds' => actuals.map { |element| element['kind'] }.uniq
      }
    end
  end
end

result = {
  'version' => 1,
  'generated_at' => Time.now.utc.iso8601,
  'status' => control_blockers.empty? && kind_blockers.empty? ? 'PASS' : 'FAIL',
  'controls_census' => control_census_path && File.basename(control_census_path),
  'unresolved_controls' => control_blockers,
  'renamed_chart_family_mismatches' => kind_blockers
}
out_path = File.join(workdir, 'reconstruction-integrity.json')
File.write(out_path, JSON.pretty_generate(result) + "\n")

if control_blockers.any?
  warn "[FAIL] reconstruction integrity: #{control_blockers.length} control(s) remain unfinished:"
  control_blockers.each do |blocker|
    warn "       - #{blocker['kind']}:#{blocker['name']} (#{blocker['status']})"
  end
  warn '       Complete the wiring/materialization, or add a reasoned controls-waivers.json entry.'
end
if kind_blockers.any?
  warn "[FAIL] reconstruction integrity: #{kind_blockers.length} renamed chart(s) changed family without residue:"
  kind_blockers.each do |blocker|
    warn "       - #{blocker['source_tile'].inspect} -> #{blocker['renamed_element'].inspect}: " \
         "expected #{blocker['expected_family']}, built #{blocker['built_families'].join('/')}"
  end
  warn '       Restore the source family, or add a reasoned png-read.json kind_waivers entry.'
end

if result['status'] == 'PASS'
  puts "[OK] reconstruction integrity: controls terminal, renamed chart families faithful or explicitly waived"
  puts "wrote #{out_path}"
  exit 0
end
warn "wrote #{out_path}"
exit 1
